lib/hooks/use-product-results.ts

```diff
@@ -0,0 +1,62 @@
+import { SerializeFrom } from '@remix-run/node';
+import { useEffect, useState } from 'react';
+import { getEcomApi, IProductFilters, Product, ProductSortBy, Slice } from '../ecom';
+import { getErrorMessage } from '../utils';
+
+export interface UseProductResultsArgs {
+    categorySlug: string;
+    filters: IProductFilters;
+    sorting: ProductSortBy;
+    initialResultsFromLoader: Slice<Product | SerializeFrom<Product>>;
+}
+
+/**
+ * Manages product slicing and loads next slices from the client side
+ * without a full page reload. Syncs with initialResultsFromLoader.
+ */
+export function useProductResults({
+    categorySlug,
+    filters,
+    sorting,
+    initialResultsFromLoader,
+}: UseProductResultsArgs) {
+    const [productSlice, setProductSlice] = useState(initialResultsFromLoader);
+    useEffect(() => {
```

Comment:
`initialResultsFromLoader` will change when the filters or the category has been changed, and the route returned the first batch of results, and we need to discard the subsequent batches that we loaded client-side.

Reply:
Maybe there are better options, but here's one:

1. Have a `const counter = useRef(0);`
2. In `dataFromLoader` effect: `counter++`
3. In `onUnmount` effect: `counter++`
4. In `loadMore`: read the counter before making the request, after the request has completed, compare the current value of the counter to the previous. If they don't match, ignore the result of the request.

Reply:
fixed, also I think name `resultsFromLoaderVersionRef` makes more sense for this counter